### DISCUSSION TITLE
How SSH works in Site-to-Site VPN over the active firewall? and pls  provide some helpfull documents or modules to get the idea of use of PSK in IPsec VPN connection.

### DISCUSSION THREAD

**Discussion by satanu01** [strongswan/strongswan#517] [Q&A]
We have successfully established Strongswan Site-to Site VPN between two networks. But we are facing two issues.
The issues are,

- First, SSH is working only when we disable the firewall in both the sites. We tried adding some rules to allow SSH, but firewall is still blocking them. So how SSH can work when the firewall is active?

- Second, if possible could you please provide some helpful documents or modules to get the idea of use of PSK in IPsec VPN connection. Means, how the authentication and encryption key is generated from the PSK? 
  A sample scenario is given below-

  ```
  pi@raspberrypi:~ $ sudo ip xfrm state
  src 10.3.56.103 dst 10.1.35.249
  	proto esp spi 0xcf9d3a19 reqid 1 mode tunnel
  	replay-window 0 flag af-unspec
  	auth-trunc hmac(sha1) 0x17a474bdee1f930f7cb6ec8894254d546f509d46 96      <------------------------      (These Keys)
  	enc cbc(aes) 0x3bcc0206907b31775c2b40d6a6499301cc7376f863933b6f38ae99c293b3ee50   <------------------------ 
  	anti-replay context: seq 0x0, oseq 0x1c2, bitmap 0x00000000
  src 10.1.35.249 dst 10.3.56.103
  	proto esp spi 0xc621c4b4 reqid 1 mode tunnel
  	replay-window 32 flag af-unspec
  	auth-trunc hmac(sha1) 0xdcd9734fd4cf64dee028efff0b167cbc6e663dbe 96      <------------------------ 
  	enc cbc(aes) 0x680a5f5f6e80a94305370cb27b4e074dfe013369b994752c9b12e899783d03cc   <------------------------ 
  	anti-replay context: seq 0x1a5, oseq 0x0, bitmap 0xffffffff
  ```

  ```
  -------------------------------------------------ipsec.secrets---------------------------------------------------------------
  -----------------------------------------------------------------------------------------------------------------------------
  # This file holds shared secrets or RSA private keys for authentication.
  10.3.56.103 10.1.35.249 : PSK "PmcNwSgSe2Luv5E6N9Szu51sCNYXuP44"        <------------------------ (From this PSK)
  
  # RSA private key for this host, authenticating it to any other host
  # which knows the public part.
  
  # this file is managed with debconf and will contain the automatically created private key
  include /var/lib/strongswan/ipsec.secrets.inc
  ```

**Comment by tobiasbrunner** [comment]
>     * We tried adding some rules to allow SSH, but firewall is still blocking them. So how SSH can work when the firewall is active?

Please add more information to your question (e.g. your config, the status output, the policies, the firewall rules, what hosts you try to reach from what hosts via SSH etc., see #196).

>     * Means, how the authentication and encryption key is generated from the PSK?

They aren't. The PSK is only used to authenticate the peers, not to derive any keys (with IKEv2, IKEv1 did that differently). The actual keys are derived from an ephemeral DH exchange. Refer to [RFC 7296](https://datatracker.ietf.org/doc/html/rfc7296) for details.

**Reply by satanu01** [reply]
First of all, Thank You for your response and guidance......

Our set up like this -

​          **Client_1                                     Gateway_1                                        Gateway_2                                     Client_2**

​                                                        `10.3.56.69 (wlan0) <--------> 10.3.63.188 (wlan0)`               

`192.168.1.101 (eth0) <----> 192.168.1.1 (eth0)           192.168.2.1 (eth0) <---->  192.168.2.111 (eth0)`

Firstly, we have followed this [link](https://www.tecmint.com/setup-ipsec-vpn-with-strongswan-on-debian-ubuntu/) for setup a Site-to-Site IPsec VPN. The content of `ipsec.config` file (Gateway_1) is -

```
config setup
        charondebug="all"
        uniqueids=yes
conn gateway1-to-gateway2
        type=tunnel
        auto=route
        authby=secret
        left=10.3.56.69
        leftsubnet=192.168.1.101/24
        right=10.3.63.188
        rightsubnet=192.168.2.111/24
        ike=aes256-sha1-modp1024!
        esp=aes256-sha1!
        aggressive=no
        keyingtries=%forever
        ikelifetime=28800s
        lifetime=3600s
        dpddelay=30s
        dpdtimeout=120s
        dpdaction=restart
```

After the VPN establishment between the two site, we can `ping` each other successfully but when we use `ssh` or `scp` for remote access these can not be working, it stuck at the terminal. So we use a command: `sudo ufw allow ssh` , it was added two rules for `ssh`, but its not working. Then we disable the firewall by `sudo ufw disable`  and bingo! the `ssh` and `scp`  both are working now. Policy details is given below-
```
pi@raspberrypi:~ $ sudo ip xfrm policy
src 192.168.1.0/24 dst 192.168.2.0/24 
	dir out priority 375423 
	tmpl src 10.3.56.69 dst 10.3.63.188
		proto esp spi 0xcf9d3a19 reqid 1 mode tunnel
src 192.168.2.0/24 dst 192.168.1.0/24 
	dir fwd priority 375423 
	tmpl src 10.3.63.188 dst 10.3.56.69
		proto esp reqid 1 mode tunnel
src 192.168.2.0/24 dst 192.168.1.0/24 
	dir in priority 375423 
	tmpl src 10.3.63.188 dst 10.3.56.69
		proto esp reqid 1 mode tunnel
src 0.0.0.0/0 dst 0.0.0.0/0 
	socket in priority 0 
src 0.0.0.0/0 dst 0.0.0.0/0 
	socket out priority 0 
src 0.0.0.0/0 dst 0.0.0.0/0 
	socket in priority 0 
src 0.0.0.0/0 dst 0.0.0.0/0 
	socket out priority 0 
src ::/0 dst ::/0 
	socket in priority 0 
src ::/0 dst ::/0 
	socket out priority 0 
src ::/0 dst ::/0 
	socket in priority 0 
src ::/0 dst ::/0 
	socket out priority 0
```
But, we don't want to disable the firewall in this scenario.  So, Pls! help us that how we use  `ssh` and `scp`  over VPN when the firewall is active.

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Lastly, you have said that `"The PSK is only used to authenticate the peers .."`, so, can you please provide some helpful documents or modules that how PSK is used in Site-to-Site IPsec VPN authentication.

**Reply by tobiasbrunner** [reply]
> After the VPN establishment between the two site, we can `ping` each other successfully but when we use `ssh` or `scp` for remote access these can not be working, it stuck at the terminal.

That's always between `192.168.1.101` and `192.168.2.111`, or what IP addresses are involved in these attempts? Are these IPs on the same hosts as `192.168.1.1` and `192.168.2.1`, respectively?

> `sudo ufw allow ssh` , it was added two rules for `ssh`, b

What rules does it add? They might not cover traffic that comes out of the tunnel for some reason. In particular if the traffic is actually forwarded to another host.

To allow all traffic from the VPN tunnel, you can try adding `leftfirewall=yes` and `lefthostaccess=yes` to your configs (the latter only if you want to access the VPN gateways themselves, not hosts behind them).
